api/store.proto

```diff
@@ -0,0 +1,185 @@
+	// UpdateObjects atomically commits the supplied objects to the data store.
+	// - Returns `FailedPrecondition` and does not commit any changes if
+	//   any of the objects have changed since they were retrieved.
+	rpc UpdateObjects(UpdateObjectsRequest) returns (UpdateObjectsResponse) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// Watch starts a stream that returns any changes to objects that match
+	// the specified selectors. When the stream begins, it immediately sends
+	// an empty message back to the client. It is important to wait for
+	// this message before taking any actions that depend on an established
+	// stream of changes for consistency.
+	rpc Watch(WatchRequest) returns (stream WatchMessage) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// View creates a snapshot of the current state of the store, and
+	// accepts a stream of queries against it. Each query returns a
+	// response in the response stream.
+	// TODO(aaronl): Does it make sense to support this in the API?
+	//rpc View(stream ViewRequest) returns (stream ViewResponse) {
+	//	option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	//}
+}
+
+message GetObjectRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// ID of the object to retrieve.
+	string object_id = 2 [(gogoproto.customname) = "ObjectID"];
+}
+
+message FindObjectsRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// Selectors are combined with OR logic. An empty list of selectors
+	// causes all objects of the given type to be returned.
+	repeated SelectBy selectors = 2;
+}
+
+message FindObjectsResponse {
+	repeated Object results = 1;
+}
+
+message UpdateObjectsRequest {
+	repeated Object objects = 1;
+}
+
+message UpdateObjectsResponse {
+	// Objects contains the updated objects. On a successful return, these
+	// will be the same as the objects passed in UpdateObjectsRequest,
+	// except with Meta updated.
+	repeated Object objects = 1;
+
+	// TODO(aaronl): Provide structured information about which object
+	// couldn't be updated, and why, in error cases?
+}
+
+message WatchRequest {
+	message WatchEntry {
+		// Kind can contain a builtin type such as "node", "secret", etc. or
+		// the kind specified by a custom-defined object.
+		string kind = 1;
+
+		// Action (create/update/delete)
+		StoreActionKind action = 2;
+
+		// Filters are combined using AND logic - an event must match
+		// all of them to pass the filter.
+		repeated SelectBy filters = 3;
+	}
+
+	// Multiple entries are combined using OR logic - i.e. if an event
+	// matches all of the selectors specified in any single watch entry,
+	// the event will be sent to the client.
+	repeated WatchEntry entries = 1;
```

Comment:
Do you think that making `action` inside `WatchEntry` optional or repeated would solve this? Then you would need a `WatchEntry` for each swarm object type, which seems reasonable. Are there other changes you have in mind?